### PR TITLE
Device Groups

### PR DESCRIPTION
**Pull request by pcdiem**
## Description:

The device groups module provides a framework to allow multiple devices to be in a group with values such as power, light color/temperature/brightness, PWM values, sensor values, etc. shared with other devices in the group. For example, with multiple light modules in a device group, the light settings can be changed on one module and the settings will automatically be changed on the other light modules. Dimmer switch modules could be in a device group with light modules and the dimmer switch could control the power, brightness and colors of all the lights in the group. Multiple dimmer switches could be in a device group to form a 3-way/4-way dimmer switch.

UDP broadcasts, followed by UDP unicasts if necessary, are used to send updates to all devices so updates are fast. There is no need for an MQTT server but all the devices in a group must be on the same IP network.

To include device groups support in the build, define USE_DEVICE_GROUPS in your user_config_override. This adds 3.5K to the code size. All devices in a group must be running firmware with device group support and have device groups enabled.

## Checklist:
  - [x] The pull request is done against the latest dev branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR.
  - [x] The code change is tested and works on core 2.6.1
  - [x] The code change pass travis tests. **Your PR cannot be merged unless tests pass**
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

### CODE REVIEW

File: Device_Groups.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+UDP broadcasts, followed by UDP unicasts if necessary, are used to send updates to all devices so updates are fast. There is no need for an MQTT server but all the devices in a group must be on the same IP network.
+
+To include device groups support in the build, define USE_DEVICE_GROUPS in your user_config_override. This adds 3.5K to the code size. All devices in a group must be running firmware with device group support and have device groups enabled.
+
+To enable device groups, set Option16 to 1 and **restart the device**.
```
Comment (s-hadinger) [inline Device_Groups.md]: Why use SetOption16? Isn't it better to create a new SetOption and trigger an automatic restart when it is changed?

File: tasmota/settings.h [hunk 2]
```diff
@@ -573,4 +573,9 @@ typedef union {
   ADC_MODE(ADC_VCC);                       // Set ADC input for Power Supply Voltage usage
 #endif
 
+// Settings re-purposed for device groups
+#define device_groups_enabled  ws_clock_reverse   // SetOption16 - Enable device groups
+#define device_group_share_in  domoticz_sensor_idx[0]    // Bitmask of device group items imported
+#define device_group_share_out domoticz_sensor_idx[1]    // Bitmask of device group items exported
```
Comment (s-hadinger) [inline tasmota/settings.h]: Theo, is it ok to re-purpose those parameters, or should we assign a fresh new location?

File: tasmota/support_device_groups.ino [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+void SendDeviceGroupPacket(IPAddress ip, char * packet, int len, const char * label)
+{
+  for (int attempt = 1; attempt <= 5; attempt++) {
+    if (PortUdp.beginPacket(ip, 1900)) {
+      PortUdp.write(packet, len);
+      if (PortUdp.endPacket()) return;
+    }
+    delay(10);
+  }
+  AddLog_P2(LOG_LEVEL_ERROR, "DGR: error sending %s packet", label);
```
Comment (s-hadinger) [inline tasmota/support_device_groups.ino]: Please use PSTR to place string literals in Flash:

AddLog_P2(LOG_LEVEL_ERROR, PSTR("DGR: error sending %s packet"), label);

File: tasmota/support_device_groups.ino [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+  if (message_type == DGR_MSGTYP_FULL_STATUS) {
+
+    // Set the flag indicating we're currently building a status message. SendDeviceGroupMessage
+    // will build but not send messages while this flag is set.
+    building_status_message = true;
+
+    // Call the drivers to build the status update.
+    if (!++outgoing_sequence) outgoing_sequence = 1;
+#ifdef DEVICE_GROUPS_DEBUG
+    AddLog_P2(LOG_LEVEL_DEBUG, "Building device group %s full status packet", device_group->group_name);
```
Comment (s-hadinger) [inline tasmota/support_device_groups.ino]: AddLog_P2(LOG_LEVEL_DEBUG, PSTR("Building device group %s full status packet"), device_group->group_name);

File: tasmota/support_device_groups.ino [hunk 5]
```diff
@@ -0,0 +1,715 @@
...
+
+  else {
+    bool shared;
+    uint8_t item;
+    uint32_t value;
+    uint8_t * value_ptr;
+    va_list ap;
+
+#ifdef DEVICE_GROUPS_DEBUG
+    AddLog_P2(LOG_LEVEL_DEBUG, "Building device group %s packet", device_group->group_name);
```
Comment (s-hadinger) [inline tasmota/support_device_groups.ino]: AddLog_P2(LOG_LEVEL_DEBUG, PSTR("Building device group %s packet"), device_group->group_name);

File: tasmota/settings.h (hunk 2 above)
Comment (arendst) [reply tasmota/settings.h]: Not OK. It removes Domoticz functionality.

As you suggested, let's reserver a new setoption and two uint16_t in Settings area.

File: tasmota/support_device_groups.ino [hunk 6]
```diff
@@ -0,0 +1,715 @@
...
+    *message_ptr++ = outgoing_sequence & 0xff;
+    *message_ptr++ = outgoing_sequence >> 8;
+
+    value = 0;
+    if (message_type == DGR_MSGTYP_UPDATE_MORE_TO_COME)
+      value |= DGR_FLAG_MORE_TO_COME;
+    else if (message_type == DGR_MSGTYP_UPDATE_DIRECT)
+      value |= DGR_FLAG_DIRECT;
+#ifdef DEVICE_GROUPS_DEBUG
+    AddLog_P2(LOG_LEVEL_DEBUG, ">sequence=%u, flags=%u", outgoing_sequence, value);
```
Comment (s-hadinger) [inline tasmota/support_device_groups.ino]: Itoo

File: tasmota/support_device_groups.ino [hunk 7]
```diff
@@ -0,0 +1,715 @@
...
+                previous_message_ptr += 5;
+                message_ptr += 5;
+                break;
+            }
+          }
+          kept_item_count++;
+        }
+      }
+#ifdef DEVICE_GROUPS_DEBUG
+    AddLog_P2(LOG_LEVEL_DEBUG, "%u items carried over from previous update", kept_item_count);
```
Comment (s-hadinger) [inline tasmota/support_device_groups.ino]: Same for all logs...

File: tasmota/settings.h (hunk 2 above)
Comment (arendst) [reply tasmota/settings.h]: Do we need Device Groups? Does adding this flood UDP traffic?

**Comment by arendst** [comment]
Luckily this one is closed by Paul...

File: Device_Groups.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
+# Device Groups
+
+The device groups module provides a framework to allow multiple devices to be in a group with values such as power, light color/temperature/brightness, PWM values, sensor values, etc. shared with other devices in the group. For example, with multiple light modules in a device group, the light settings can be changed on one module and the settings will automatically be changed on the other light modules. Dimmer switch modules could be in a device group with light modules and the dimmer switch could control the power, brightness and colors of all the lights in the group. Multiple dimmer switches could be in a device group to form a 3-way/4-way dimmer switch.
+
+UDP broadcasts, followed by UDP unicasts if necessary, are used to send updates to all devices so updates are fast. There is no need for an MQTT server but all the devices in a group must be on the same IP network.
```
Comment (s-hadinger) [inline Device_Groups.md]: Can you please explain when it sends broadcast and when it needs unicast?

**Comment by pcdiem** [comment]
I just re-purposed existing settings and options as an initial step 
during development to prevent conflicts with other potential development 
changes. I assumed that if the updates are accepted, we'd assign 
dedicated settings and options.

File: Device_Groups.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+The items that are sent to the group and the items that are received from the group are selected with the DevGroupShare command. By default all items are sent and received from the group. An example of when the DevGroupShare command would be used is when you have a group of lights that you control with a dimmer switch and home automation software. You want the dimmer switch to be able to control all items. The home automation software controls each light individually. When it controls the whole group, it actually sends command to each light in the group. If you use the home automation software to turn an individual light on or off or change it’s brightness, color or scheme, you do not want the change to be replicated to the other lights. In this case, you would set the incoming and outgoing item masks to 255 on the dimmer switch (DevGroupShare 255,255) and set the incoming item mask to 255 and outgoing item mask to 0 on all the lights (DevGroupShare 255,0).
```
Comment (s-hadinger) [inline Device_Groups.md]: What are the other values possible for masks?

**Comment by arendst** [comment]
I'll reserve them for you now. Hold on.

**Review by s-hadinger — COMMENTED** [review]
This looks interesting, I'm looking forward to trying it.

**Comment by arendst** [comment]
How many bits do you need for share in and share out? is 16 enough or do you need 32?

File: tasmota/settings.h (hunk 2 above)
Comment (s-hadinger) [reply tasmota/settings.h]: That's why I'm curious about broadcast/unicast packets

**Comment by pcdiem** [comment]
It sends one broadcast for each update. If it does not receive an ack 
from any of the devices in the group, it follows up with unicasts to 
those devices.

The point of the broadcast it to get the update to all the devices in 
the group in unison as much as possible. So, for example, as light 
brightness or colors are changed, they change on all light simultaneously.

**Comment by pcdiem** [comment]
That depends on how adopted this will be. I initially included support 
in the light module and the PWM DImmer module I will submit a pull 
request for today. The idea is to have most related items grouped 
together into one share in/out bit. 32 would, of course, allow for more 
growth in we can afford the space.

**Comment by arendst** [comment]
Then 32 it will be.

**Comment by pcdiem** [comment]
I had to close the request because I failed to test with minimal builds. 
The module uses the same multicast code from support_udp so it requires 
that USE_EMULATION be defined. I added undefs for the lighter builds and 
will re-pull after the build is tested.

**Comment by s-hadinger** [comment]
Thanks, so Broadcast traffic should be minimal: 1 UDP packet per command.

**Comment by pcdiem** [comment]
One UDP broadcast per command/update. After that, UDP unicasts for acks 
to the originator and UDP unicasts to any devices that didn't ack the 
broadcast.

**Comment by arendst** [comment]
Using latest dev use SetOption85 for Device Groups control and 
```
  uint32_t      device_group_share_in;     // FCC - Bitmask of device group items imported
  uint32_t      device_group_share_out;    // FD0 - Bitmask of device group items exported
```
in Settings

**Comment by pcdiem** [comment]
Thanks. I'll modify to use those settings and use PSTR for the string 
constants. I'll re-submit in a few hours.

**Pull request closed without merging**